### DISCUSSION TITLE
1.7.2

### DISCUSSION THREAD

**Discussion by elrido** [PrivateBin/PrivateBin#1308] [Announcements]
* ADDED: Allow use of `shortenviayourls` in query parameters (#1267)
* ADDED: Input sanitation to some not yet filtered query and server parameters
* ADDED: Optional Bootstrap CSS 5.3.3 based template, use configuration `template = "bootstrap5"` to switch to it (#728)
* CHANGED: "Send" button now labeled "Create" (#946)
* CHANGED: Drop some PHP < 5.6 fallbacks, minimum version is PHP 7.3 as of release 1.6.0
* CHANGED: Set `lang` cookie with lax `SameSite` property
* CHANGED: Upgrading libraries to: DOMpurify 3.1.2 (#1299) & jQuery 3.7.1
* CHANGED: `create` attribute is no longer returned in API for pastes & can be disabled for comments using `discussiondatedisplay` as well (#1290)
* FIXED: Add cache control headers also to API calls (#1263)
* FIXED: Shortened paste URL does not appear in email (#606)

Note regarding the new template "bootstrap5", that if you want the button icons (SVG) to display, you have to relax the CSP rule slightly and change default-src from `'none'` to `'self'`. You configure it as follows:

```ini
template = "bootstrap5"
cspheader = "default-src 'self'; base-uri 'self'; form-action 'none'; manifest-src 'self'; connect-src * blob:; script-src 'self' 'unsafe-eval'; style-src 'self'; font-src 'self'; frame-ancestors 'none'; img-src 'self' data: blob:; media-src blob:; object-src blob:; sandbox allow-same-origin allow-scripts allow-forms allow-popups allow-modals allow-downloads"
```

The new theme comes in only one flavour, but does include a dark-mode switch and attempts detecting the currently set browser preference. 

We don't yet enable this new template by default. Please report any issues you find with it or submit pull requests with your improvements. Should no major issues get detected, we intend to make it the new default later this year and eventually deprecate and remove the old bootstrap 3 templates, as well as the page (classic ZereBin) one. It would be appreciated if additional templates would get submitted and shared with the community, so we get some more variety to choose from.

<hr /><em>This discussion was created from the release <a href='https://github.com/PrivateBin/PrivateBin/releases/tag/1.7.2'>1.7.2</a>.</em>

**Comment by proton-ab** [comment]
Some feedback about `bootstrap5`:

- We are not using full width of screen anymore
- There is no border or padding around `<pre>` element and padding for line numbers is insufficient (they seem cut off)
- `PrivateBin` name on footer is too big
- Navbar header does not use `align-items` for flexbox causing `New` button as well as `<input>` boxes to be misaligned compared to text
- `<pre>` element inherits `#fff` color from background making it very jarring to read due to contrast difference between lines

Here are my personal modifications with solve all of the issues above:

```diff
--- <old>	Sun May  5 20:52:29 2024
+++ <new>	Mon May  6 01:46:06 2024
@@ -159,48 +159,48 @@
 					<span class="navbar-toggler-icon"></span>
 				</button>
 				<div id="navbar" class="collapse navbar-collapse">
-					<ul class="navbar-nav me-auto">
+					<ul class="navbar-nav me-auto gap-2 align-items-lg-center align-items-stretch">
 						<li id="loadingindicator" class="navbar-text hidden me-auto">
 							<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#clock" /></svg>
 							<?php echo I18n::_('Loading…'), PHP_EOL; ?>
 						</li>
-						<li class="nav-item me-2">
-							<button id="retrybutton" type="button" class="reloadlink hidden btn btn-primary">
+						<li class="nav-item d-flex flex-lg-row flex-column">
+							<button id="retrybutton" type="button" class="reloadlink hidden btn btn-primary flex-fill">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#repeat" /></svg> <?php echo I18n::_('Retry'), PHP_EOL; ?>
 							</button>
 						</li>
-						<li class="nav-item">
-							<button id="newbutton" type="button" class="hidden btn btn-secondary me-2">
+						<li class="nav-item d-flex gap-2 flex-lg-row flex-column">
+							<button id="newbutton" type="button" class="hidden btn btn-secondary flex-fill">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#file-earmark" /></svg> <?php echo I18n::_('New'), PHP_EOL; ?>
 							</button>
-							<button id="clonebutton" type="button" class="hidden btn btn-secondary me-2">
+							<button id="clonebutton" type="button" class="hidden btn btn-secondary flex-fill">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#copy" /></svg> <?php echo I18n::_('Clone'), PHP_EOL; ?>
 							</button>
-							<button id="rawtextbutton" type="button" class="hidden btn btn-secondary me-2">
+							<button id="rawtextbutton" type="button" class="hidden btn btn-secondary flex-fill">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#filetype-txt" /></svg> <?php echo I18n::_('Raw text'), PHP_EOL; ?>
 							</button>
-							<button id="downloadtextbutton" type="button" class="hidden btn btn-secondary me-2">
+							<button id="downloadtextbutton" type="button" class="hidden btn btn-secondary flex-fill">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#download" /></svg> <?php echo I18n::_('Save paste'), PHP_EOL; ?>
 							</button>
 <?php
 if ($EMAIL) :
 ?>
 
-							<button id="emaillink" type="button" class="hidden btn btn-secondary me-2">
+							<button id="emaillink" type="button" class="hidden btn btn-secondary">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#envelope" /></svg> <?php echo I18n::_('Email'), PHP_EOL; ?>
 							</button>
 <?php
 endif;
 if ($QRCODE) :
 ?>
-							<button id="qrcodelink" type="button" data-toggle="modal" data-target="#qrcodemodal" class="hidden btn btn-secondary me-2">
+							<button id="qrcodelink" type="button" data-toggle="modal" data-target="#qrcodemodal" class="hidden btn btn-secondary">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#qr-code" /></svg> <?php echo I18n::_('QR code'), PHP_EOL; ?>
 							</button>
 <?php
 endif;
 ?>
 						</li>
-						<li id="expiration" class="nav-item me-2 d-flex hidden">
+						<li id="expiration" class="nav-item d-flex hidden">
 							<label for="pasteExpiration" class="form-label my-auto me-1"><?php echo I18n::_('Expires'); ?>:</label>
 							<select id="pasteExpiration" name="pasteExpiration" class="form-select">
 <?php
@@ -216,7 +216,7 @@
 ?>
 							</select>
 						</li>
-						<li class="nav-item me-2">
+						<li class="nav-item">
 							<div id="burnafterreadingoption" class="navbar-text form-check hidden">
 								<input class="form-check-input" type="checkbox" id="burnafterreading" name="burnafterreading"<?php
 if ($BURNAFTERREADINGSELECTED) :
@@ -231,7 +231,7 @@
 <?php
 if ($DISCUSSION) :
 ?>
-						<li class="nav-item me-2">
+						<li class="nav-item">
 							<div id="opendiscussionoption" class="navbar-text form-check hidden">
 								<input class="form-check-input" type="checkbox" id="opendiscussion" name="opendiscussion"<?php
 	if ($OPENDISCUSSION) :
@@ -247,8 +247,8 @@
 endif;
 if ($PASSWORD) :
 ?>
-						<li class="nav-item me-2">
-							<div id="password" class="navbar-form hidden">
+						<li class="nav-item">
+							<div id="password" class="navbar-form hidden d-flex">
 								<input type="password" id="passwordinput" placeholder="<?php echo I18n::_('Password (recommended)'); ?>" class="form-control" size="23" />
 							</div>
 						</li>
@@ -256,7 +256,7 @@
 endif;
 if ($FILEUPLOAD) :
 ?>
-						<li id="attach" class="nav-item hidden dropdown me-2">
+						<li id="attach" class="nav-item hidden dropdown d-flex">
 							<a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false"><?php echo I18n::_('Attach a file'); ?></a>
 							<ul class="dropdown-menu">
 								<li id="filewrap">
@@ -293,17 +293,11 @@
 							</select>
 						</li>
 					</ul>
-					<ul class="navbar-nav">
-						<li class="nav-item">
-							<div class="form-check form-switch navbar-text">
-								<input id="bd-theme" type="checkbox" class="form-check-input">
-								<label for="bd-theme" class="form-check-label"><?php echo I18n::_('Dark Mode'); ?></label>
-							</div>
-						</li>
+					<ul class="navbar-nav flex-row-reverse gap-2">
 <?php
 if (!empty($LANGUAGESELECTION)) :
 ?>
-						<li id="language" class="nav-item dropdown ms-2">
+						<li id="language" class="nav-item dropdown">
 							<a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false">
 								<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#flag" /></svg> <?php echo $LANGUAGES[$LANGUAGESELECTION][0], PHP_EOL; ?>
 							</a>
@@ -324,12 +318,18 @@
 <?php
 endif;
 ?>
+						<li class="nav-item">
+							<div class="form-check form-switch navbar-text">
+								<input id="bd-theme" type="checkbox" class="form-check-input">
+								<label for="bd-theme" class="form-check-label"><?php echo I18n::_('Dark Mode'); ?></label>
+							</div>
+						</li>
 					</ul>
 				</div>
 			</div>
 		</nav>
 		<main>
-			<section class="container mt-2">
+			<section class="container-fluid mt-2">
 <?php
 if (!empty($NOTICE)) :
 ?>
@@ -422,7 +422,7 @@
 					</li>
 				</ul>
 			</section>
-			<section class="container">
+			<section class="container-fluid">
 				<article class="row">
 					<div id="placeholder" class="col-md-12 hidden"><?php echo I18n::_('+++ no paste text +++'); ?></div>
 					<div id="attachmentPreview" class="col-md-12 text-center hidden"></div>
@@ -433,22 +433,22 @@
 					<p class="col-md-12"><textarea id="message" name="message" cols="80" rows="25" class="form-control hidden"></textarea></p>
 				</article>
 			</section>
-			<section class="container">
+			<section class="container-fluid">
 				<div id="discussion" class="hidden">
-					<h4><?php echo I18n::_('Discussion'); ?></h4>
+					<h5><?php echo I18n::_('Discussion'); ?></h5>
 					<div id="commentcontainer"></div>
 				</div>
 			</section>
-			<section class="container">
+			<section class="container-fluid">
 				<div id="noscript" role="alert" class="alert alert-info noscript-hide">
 					<svg width="16" height="16" fill="currentColor" aria-hidden="true"><use href="img/bootstrap-icons.svg#exclamation-circle" /></svg>
 					<?php echo I18n::_('Loading…'); ?><br />
 					<span class="small"><?php echo I18n::_('In case this message never disappears please have a look at <a href="%s">this FAQ for information to troubleshoot</a>.', 'https://github.com/PrivateBin/PrivateBin/wiki/FAQ#why-does-the-loading-message-not-go-away'); ?></span>
 				</div>
 			</section>
-			<footer class="container">
+			<footer class="container-fluid">
 				<div class="row">
-					<h4 class="col-md-5 col-xs-8"><?php echo I18n::_($NAME); ?> <small>- <?php echo I18n::_('Because ignorance is bliss'); ?></small></h4>
+					<h5 class="col-md-5 col-xs-8"><?php echo I18n::_($NAME); ?> <small>- <?php echo I18n::_('Because ignorance is bliss'); ?></small></h5>
 					<p class="col-md-1 col-xs-4 text-center"><?php echo $VERSION; ?></p>
 					<p id="aboutbox" class="col-md-6 col-xs-12">
 						<?php echo sprintf(
```

```diff
--- <old>	Sun May  5 20:52:29 2024
+++ <new>	Sun May  5 23:01:47 2024
@@ -67,3 +67,19 @@
 html[dir="rtl"] #deletelink, html[dir="rtl"] #qrcodemodalClose {
 	float: left;
 }
+
+main {
+	padding-left: 4ch;
+	padding-right: 4ch;
+}
+
+pre {
+	background-color: #f5f5f5;
+	border: 1px solid #ccc;
+	border-radius: 4px;
+	padding: 5px;
+}
+
+ol.linenums {
+	padding-left: revert;
+}
```

**Reply by proton-ab** [reply]
![chrome_2024-05-05_23-13-35](https://github.com/PrivateBin/PrivateBin/assets/25139420/01d5fc68-01cb-4bcb-9357-a9a8ffd92665)

![chrome_2024-05-05_23-15-04](https://github.com/PrivateBin/PrivateBin/assets/25139420/9eafe486-d578-4553-aaf3-2d974d660c4c)

**Reply by proton-ab** [reply]
~~There is also a known issue with my change of addition of `align-items-center` which breaks on mobile when flexbox is converted from row to column. For now I am unsure how to best handle it without replicating media selectors that Bootstrap uses.~~

Solved this and updated diff in parent comment. Screenshots for mobile below:

![chrome_2024-05-06_01-51-44](https://github.com/PrivateBin/PrivateBin/assets/25139420/f8e23bae-d350-4ccc-9853-26ea67927615)

![chrome_2024-05-06_01-51-50](https://github.com/PrivateBin/PrivateBin/assets/25139420/d2a55357-2dd6-4120-802b-77285d509368)

**Reply by elrido** [reply]
Thank you for sharing your observations and improvement. I'll start picking these up in a branch later this week. For the pre-elements, we may now need to put them into one of these [cards](https://getbootstrap.com/docs/4.0/components/card/#card-styles) to get a similar effect as in bootstrap 3. It is also annoying when sharing markdowns with code blocks, those don't stick out anymore either, so maybe using our own CSS is better after all, as long as we take care of it also working in dark mode.

**Reply by proton-ab** [reply]
>as long as we take care of it also working in dark mode.

Indeed I have forgot about this. For border, `var(--bs-border-color)` can be used. For alternate line colors however, I am unable to find satisfactory default that would work between dark and light modes so we'll need to define our own. We can rely on `[data-bs-theme=dark]` to define new variable color in custom CSS for `bootstrap5`.

```css
pre {
	border: 1px solid var(--bs-border-color);
	border-radius: 4px;
	padding: 5px;
}

[data-bs-theme=light] pre {
	background-color: #f5f5f5;
}
[data-bs-theme=dark] pre {
	background-color: #212529;
}
```

**Reply by proton-ab** [reply]
Additionally, there seem to be some clash of styles with `sons-of-obsidian` in dark mode causing line numbers to be overly dark and hard to read:

![chrome_2024-05-06_14-55-55](https://github.com/PrivateBin/PrivateBin/assets/25139420/f3d8ac42-88b5-40c4-b7a0-69f32f36b1ff)

**Reply by rugk** [reply]
Also, as usual, feel free to create a PR (better small PRs focusing one one aspect), unless @elrido now says he has already incoperated/added some stuff.)

**Reply by elrido** [reply]
Not started yet, but good call, I'll try to raise the separate fixes as separate PR's then, so we are more flexible in discussing each and their impact. I created [a project](https://github.com/orgs/PrivateBin/projects/4/views/1) to track the different issues related to this.

**Comment by sebastianw** [comment]
With this release I can't create pastes anymore. 1.7.1 works fine, with 1.7.2 I only get an error `Error saving paste. Sorry.`. I use MySQL as backend.

**Reply by elrido** [reply]
You can manually upgrade the schema by running these queries:
```SQL
ALTER TABLE "paste" DROP COLUMN "postdate";
UPDATE "config" SET "value" = '1.7.2' WHERE "id" = 'VERSION';
```

What we would need to know to address this is a) do you get any errors running one of these queries and can you share it? b) are you using MariaDB or MySQL or another flavour and which version precisely?

It may also be a permission issue, if you haven't granted the privatebin user permissions to change the schema. This is a ~safe~ secure thing to do, but you would need to upgrade manually. Fortunately we rarely change it (last one was in 1.3).

Edit: Not allowing the privatebin database user to alter the schema is a security benefit, so generally a thing I'd recommend doing. During upgrades that could be relaxed. ~In any case, we should probably clarify this bit in the install doc.~ Actually, the [exact grants and the manual alternative](https://github.com/PrivateBin/PrivateBin/blob/master/doc/Installation.md#using-a-database-instead-of-flat-files) are explained in the doc. Maybe anyone has suggestions on rewording it, though?

**Reply by sebastianw** [reply]
I did not get any errors in the webserver logfile. I use MariaDB. The problem was probably that the privatebin user had no GRANT privileges. Dropping the column by hand fixed the problem and I have now changed permissions to include the GRANT option.

**Comment by skyzuma** [comment]
same problem here ... i created the tables new and now it works but idk why an update will "crash the database connection"

**Reply by elrido** [reply]
Please check the recommendation [above](https://github.com/orgs/PrivateBin/discussions/1308#discussioncomment-9333336) and please also comment there regarding exact errors logged and exact RDBM and version used.

**Comment by jerrinss5** [comment]
Same problem I use S3 as a backend. Works with 1.7.1 but not with 1.7.2
Got the following error and created an issue for the same here: https://github.com/PrivateBin/PrivateBin/issues/1315
```
2024/05/06 16:01:58 [error] 18#18: *7961 FastCGI sent in stderr: "PrivateBin\Controller->_create()
#8 /var/www/index.php(18): PrivateBin\Controller->__construct()
#9 {main}
  thrown in /srv/vendor/aws/aws-sdk-php/src/Credentials/CredentialProvider.php on line 72" while reading response header from upstream, client: <CUT>, server: , request: "POST / HTTP/1.1", upstream: "fastcgi://unix:/run/php-fpm.sock:", host: "<CUT>"
```

**Reply by elrido** [reply]
Check out https://github.com/PrivateBin/docker-nginx-fpm-alpine/issues/192 - we found this to be an issue in the main container image (wrong version of the AWS SDK gets selected by composer) and for now you can switch to the `edge` tag where the issue didn't occur, until we release new image tags with a fixed setup.

**Comment by rugk** [comment]
****Important:** Please _don't_ update to this release, it contains a critical bug! See https://github.com/PrivateBin/PrivateBin/issues/1309 for details.**
